lib/CliftTransforms/LoopDetection.cpp

```diff
@@ -0,0 +1,211 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <ranges>
+
+#include "llvm/ADT/MapVector.h"
+
+#include "mlir/Pass/Pass.h"
+
+#include "revng/Clift/Clift.h"
+#include "revng/Clift/CliftOpHelpers.h"
+#include "revng/CliftTransforms/Passes.h"
+
+namespace mlir {
+namespace clift {
+#define GEN_PASS_DEF_CLIFTLOOPDETECTION
+#include "revng/CliftTransforms/Passes.h.inc"
+} // namespace clift
+} // namespace mlir
+
+namespace clift = mlir::clift;
+using namespace clift;
+
+namespace {
+
+template<typename IteratorT, typename SentinelT>
+static bool isOrderedBefore(IteratorT First, IteratorT Second, SentinelT End) {
+  for (; First != End; ++First) {
+    if (First == Second)
+      return true;
+  }
+  return false;
+}
+
+static bool isBackwardGoto(GotoOp Goto, AssignLabelOp Label) {
+  mlir::Block *LabelBlock = Label->getBlock();
+  mlir::Operation *CurrentOp = Goto.getOperation();
+
+  // Traverse up through nesting branch operations until the current operation
+  // and the target label operation are directly nested within the same block.
+  while (CurrentOp->getBlock() != LabelBlock) {
+    mlir::Operation *ParentOp = CurrentOp->getParentOp();
+
+    if (not mlir::isa<BranchOpInterface>(ParentOp))
+      return false;
+
+    CurrentOp = ParentOp;
+  }
+
+  return isOrderedBefore(Label->getIterator(),
+                         CurrentOp->getIterator(),
+                         LabelBlock->end());
+}
+
+static mlir::Block *extractAsBlock(mlir::Block *Block,
+                                   mlir::Block::iterator Begin,
+                                   mlir::Block::iterator End) {
+  mlir::Block *NewBlock = new mlir::Block();
+
+  NewBlock->getOperations().splice(NewBlock->begin(),
+                                   Block->getOperations(),
+                                   Begin,
+                                   End);
+
+  return NewBlock;
+}
+
+static mlir::Block::iterator findLoopBoundary(mlir::Block *NestingBlock,
+                                              mlir::Operation *BoundingOp) {
+  while (BoundingOp->getBlock() != NestingBlock)
+    BoundingOp = BoundingOp->getParentOp();
+
+  return std::next(BoundingOp->getIterator());
+}
+
+static mlir::Value createLabel(mlir::OpBuilder &Builder,
+                               FunctionOp Function,
+                               mlir::Location Location) {
+  mlir::OpBuilder::InsertionGuard Guard(Builder);
+  Builder.setInsertionPointToStart(&Function.getBody().front());
+  return Builder.create<MakeLabelOp>(Location);
+}
+
+static void createLoop(FunctionOp Function,
+                       AssignLabelOp LoopLabel,
+                       llvm::ArrayRef<GotoOp> Gotos) {
+  mlir::Location LoopLoc = mlir::UnknownLoc::get(Function->getContext());
+
+  mlir::Block *LabelBlock = LoopLabel->getBlock();
+  bool BoundingGotoIsInLabelScope = Gotos.back()->getBlock() == LabelBlock;
+
+  auto BodyBegin = std::next(LoopLabel->getIterator());
```

Comment:
Wait, the `LoopLabel` should stay inside the loop body according to the REP.
I realize that at this point the semantic is equivalent.
But do you have any reason to leave it out?